pkg/kritis/review/review.go

```diff
@@ -50,49 +50,130 @@ func New(c *Config) Reviewer {
+			allowlist = append(allowlist, pattern.NamePattern)
+		}
 	}
+	return allowlist
+}
+
+// reviewImageWithGAP reviews single image against a single generic attestation policy
+// It returns:
+//   - (bool) whether the image is attested
+//   - ([] string) a list of unsatisfied attestation authority names
+//   - (err) error if fails to get attestation authorities from policy
+func (r Reviewer) reviewImageWithGAP(image string, gap v1beta1.GenericAttestationPolicy, c metadata.ReadOnlyClient) (bool, []string, error) {
+	glog.Infof("Validating against GenericAttestationPolicy %s", gap.Name)
 
+	// Get all AttestationAuthorities in this policy.
+	auths, err := r.getAttestationAuthoritiesForGAP(gap)
+	if err != nil {
+		return false, nil, err
+	}
+
+	notAttestedAuthNames := r.findUnsatisfiedAuthNames(image, auths, c)
+	if len(notAttestedAuthNames) == 0 {
+		glog.Infof("Image is attested by gap %s", gap.Name)
+		return true, make([]string, 0), nil
+	}
+
+	glog.Infof("Image is missing attestations from %+q in gap %s", notAttestedAuthNames, gap.Name)
+	return false, notAttestedAuthNames, nil
+}
+
+func (r Reviewer) checkGAPs(gaps []v1beta1.GenericAttestationPolicy) error {
+	for _, gap := range gaps {
+		// Get all AttestationAuthorities in this policy.
+		auths, err := r.getAttestationAuthoritiesForGAP(gap)
+		if err != nil {
+			return err
+		}
+		if len(auths) == 0 {
+			return fmt.Errorf("Generic attestation policy %s did not specify any attestation authority.", gap.Name)
+		}
+	}
+	return nil
+}
+
+// Helper function to generate a summary of missing attestations.
+func generateMissingAttestationSummary(badImages []string, missingAttestations map[string]map[string][]string) string {
+	errMsg := "Some images are not attested by all authorities in any GAP policy:\n"
+	for _, image := range badImages {
+		errMsg = errMsg + fmt.Sprintf("- image %s is not attested by\n", image)
+		for gapName, authNames := range missingAttestations[image] {
+			errMsg = errMsg + fmt.Sprintf("  - %+q in gap %s\n", authNames, gapName)
+		}
+	}
+	return errMsg
+}
+
+// ReviewGAP reviews images against generic attestation policies
+// Return error if policies are not well-formed or violations are found.
+func (r Reviewer) ReviewGAP(images []string, gaps []v1beta1.GenericAttestationPolicy, pod *v1.Pod, c metadata.ReadOnlyClient) error {
+	// If no policy found, images should be admitted.
 	if len(gaps) == 0 {
 		glog.Info("No Generic Attestation Policies found")
 		return nil
 	}
 
+	// Check if all GAPs are well-formed.
+	if err := r.checkGAPs(gaps); err != nil {
+		return err
+	}
+
+	images, removed := util.RemoveGloballyAllowedImages(images)
+	glog.Infof("%d images are globally allowed: %+q\n", len(removed), removed)
+
+	images, removed = util.RemoveGapAllowedImages(images, generateGapAllowlist(gaps))
+	glog.Infof("%d images are further gap allowed: %+q\n", len(removed), removed)
+
+	missingAttestations := map[string]map[string][]string{}
```

Comment:
Please document at high-level what the key/values mean in this case.

Reply:
done.